# Chromatin evolution

This repository contains all the necessary code to reproduce the comparative proteomics & comparative genomics analyses of our manuscript [**Comparative proteogenomics deciphers the origin and evolution of eukaryotic chromatin**](https://github.com/sebepedroslab/chromatin-evolution-analysis) (Grau-Bov√© et al. 2021).

The data and code are structured as follows:

1. **Classification of canonical and variant histones**: see `results_histones_phylo` folder. Classification of histone variants. Also includes analysis of histone dobulets and gene clusters.

2. **Evolutionary analysis of the chromatin enzymatic machinery and readers**: see `results_toolkit_phylo` folder. Identification of orthologous groups in chromatin  modification-related enzyme families from selected eukaryotic species, analyses of their prokaryotic homology, fusions with transposon-associated domains, and more.

3. **hPTM conservation**: see `results_PTMs` folder. Includes consensus alignments and hPTM master table.

```python
 .              +   .                .   . .     .  .
                   .                    .       .     *
  .       *                        . . . .  .   .  + .
          Hic sunt dracones            .   .  +  . . .
.                 |             .  .   .    .    . .
                  |           .     .     . +.    +  .
                 \|/            .       .   . .
        . .       V          .    * . . .  .  +   .
           +      .           .   .      +
                            .       . +  .+. .
  .                      .     . + .  . .     .      .
           .      .    .     . .   . . .        ! /
      *             .    . .  +    .  .       - O -
          .     .    .  +   . .  *  .       . / |
               . + .  .  .  .. +  .
.      .  .  .  *   .  *  . +..  .            *
 .      .   . .   .   .   . .  +   .    .            +
```
